server/gitrest/src/routes/git/refs.ts

```diff
@@ -28,24 +31,66 @@ async function getRefs(repoManager: utils.RepositoryManager, owner: string, repo
     return refsP.map((ref) => refToIRef(ref));
 }
 
-async function getRef(repoManager: utils.RepositoryManager, owner: string, repo: string, refId: string): Promise<IRef> {
+async function getRef(
+    repoManager: utils.RepositoryManager,
+    owner: string,
+    repo: string,
+    refId: string,
+    externalStorageManager: ExternalStorageManager): Promise<IRef> {
     const repository = await repoManager.open(owner, repo);
-    const ref = await git.Reference.lookup(repository, refId, undefined);
-    return refToIRef(ref);
+    try {
+        const ref = await git.Reference.lookup(repository, refId, undefined);
+        return refToIRef(ref);
+    } catch (err) {
+        if (process.env.EXTERNAL_STORAGE_ENABLED != "true") {
+            winston.info(`External storage is not enabled`);
+            return;
+        }
+        winston.error(`getRef error: ` + err);
+        // Lookup external storage if commit does not exist.
+        winston.info(`Ref# Ref not found!: ${repo} : ${refId}`);
+        const fileName = refId.substring(refId.lastIndexOf("/") + 1);
+        // If file does not exist or error trying to look up commit, return the original error.
+        try {
+            await externalStorageManager.readAndSync(repo, fileName);
+            winston.info("Rehydration completed. Getting ref");
+            return getRef(repoManager, owner, repo, refId, externalStorageManager);
+        } catch (bridgeError) {
+            winston.error(`Giving up on creating ref. BridgeError: ${bridgeError}`);
+            return Promise.reject(err);
+        }
+    }
 }
 
 async function createRef(
     repoManager: utils.RepositoryManager,
     owner: string,
     repo: string,
-    createParams: ICreateRefParams): Promise<IRef> {
+    createParams: ICreateRefParamsExternal,
+    externalStorageManager: ExternalStorageManager,
+): Promise<IRef> {{
     const repository = await repoManager.open(owner, repo);
     const ref = await git.Reference.create(
         repository,
         createParams.ref,
         git.Oid.fromString(createParams.sha),
         0,
         "");
+    // tslint:disable-next-line
+    winston.info(`CREATE REF!!!! ${repo}`);
+
+    if (process.env.EXTERNAL_STORAGE_ENABLED != "true") {
+        winston.info(`External storage is not enabled`);
+    } else {
+        if (createParams.config.shouldWriteToExternalStorage) {
+            try {
+                await externalStorageManager.writeFile(repo, createParams.ref, createParams.sha, false);
+            } catch (e) {
+                // ignored
```

Comment:
Why do we not want to log this error?

Reply:
Good question - will check with Pradeep or Tanvir